/*
 * Copyright (c) 2013-2014,2016 - Adjacent Link LLC, Bridgewater,
 * New Jersey
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 *
 * * Redistributions of source code must retain the above copyright
 *   notice, this list of conditions and the following disclaimer.
 * * Redistributions in binary form must reproduce the above copyright
 *   notice, this list of conditions and the following disclaimer in
 *   the documentation and/or other materials provided with the
 *   distribution.
 * * Neither the name of Adjacent Link LLC nor the names of its
 *   contributors may be used to endorse or promote products derived
 *   from this software without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
 * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
 * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
 * FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
 * COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
 * INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
 * BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
 * LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
 * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
 * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
 * ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
 * POSSIBILITY OF SUCH DAMAGE.
 *
 * See toplevel COPYING for more information.
 */

// Transmission Protocol
//
// OpenStatistic API messages are transmitted over a TCP connection
// established between the libostatistic server and a client
// application. The libostatistic can support multiple simultaneous
// client connections. Each API message is transmitted using length
// prefix framing, where the length of the message is sent as an
// unsigned 32-bit integer value (4 bytes) in network byte order
// preceding the transmission of the serialized message. The length
// prefix framing exists outside the Google Protocol Buffers
// specification and is used to delineate message boundaries in a data
// stream.

package OpenStatisticAPI;

option optimize_for = SPEED;

// query message types
enum QueryMessageType
{
  TYPE_QUERY_STATISTIC = 1;
  TYPE_QUERY_STATISTICTABLE = 2;
}

// update message types
enum UpdateMessageType
{
  TYPE_UPDATE_STATISTICCLEAR = 1;
  TYPE_UPDATE_STATISTICTABLECLEAR = 2;
}

// Any message is used to hold any of the allowable
// configuration parameter and statistic types
message Any
{
  enum AnyType
  {
    TYPE_ANY_INT64 = 1;
    TYPE_ANY_UINT64 = 2;
    TYPE_ANY_DOUBLE = 3;
    TYPE_ANY_STRING = 4;
    TYPE_ANY_BOOLEAN = 5;
  }
  
  // Any data type
  required AnyType type = 1;
  
  // stores UINT64
  optional uint64 u64Value = 2;

  // stores INT64
  optional sint64 i64Value = 3;

  // stores Double
  optional double dValue = 4;

  // stores Boolean
  optional bool bValue = 5;

  // stores String
  optional string sValue = 6;
}

message Request
{
  message Query
  {
    // Query the emulator for component statistic values
    message Statistic
    {
      // statistic element names to retrieve
      // specify none to retrieve all
      repeated string names = 1;
    }

    // Query the emulator for component statistic table values
    message StatisticTable
    {
      // statistic table names to retrieve
      // specify none to retrieve all
      repeated string names = 1;
    }
    
    // set to the appropriate Query type
    required QueryMessageType type = 1;

    // must be present when type == TYPE_QUERY_STATISTIC
    optional Statistic statistic = 2;

    // must be present when type == TYPE_QUERY_STATISTICTABLE
    optional StatisticTable statisticTable = 3;
  }
  

  message Update
  {
    // Clear component statistic element values
    message StatisticClear
    {
      // statistic element names to clear
      // specify none to clear all
      repeated string names = 1;
    }

    // Clear component statistic table values
    message StatisticTableClear
    {
      // statistic table names to clear
      // specify none to clear all
      repeated string names = 1;
    }
    
    // set to the appropriate Update type
    required UpdateMessageType type = 1;

    // must be present when type == TYPE_UPDATE_STATISTICCLEAR
    optional StatisticClear statisticClear = 2;

    // must be present when type == TYPE_UPDATE_STATISTICTABLECLEAR
    optional StatisticTableClear statisticTableClear = 3;
  }

  // Request message types
  enum RequestMessageType
  {
    TYPE_REQUEST_QUERY = 1;
    TYPE_REQUEST_UPDATE = 2;
  }

  // message sequence number referenced in corresponding
  // response message
  required uint32 sequence = 1;
  
  // set to the appropriate Request message type
  required RequestMessageType type = 2;

  // must be present when type == TYPE_REQUEST_QUERY
  optional Query query = 3;

  // must be present when type == TYPE_REQUEST_UPDATE
  optional Update update = 4;
}

message Response
{ 
  message Query
  {
    // Query response with component statistic
    // element values
    message Statistic
    {
      message Element
      {
        // statistic element name
        required string name = 1;

        // statistic element value
        required Any value = 2;
      }
      
      // 0 or more statistic elements
      repeated Element elements = 1;
    }

    // Query response with component statistic
    // table values
    message StatisticTable
    {
      message Table
      {
        message Row
        {
          // row of table values
          repeated Any values = 1;
        }
        
        // table name
        required string name = 1;

        // table rows
        repeated Row rows = 2;

        // table column labels
        repeated string labels = 3;
      }

      // 0 or more statistic tables
      repeated Table tables = 1;
    }

    // set to the appropriate Query type
    required QueryMessageType type = 1;

    // must be present when type == TYPE_QUERY_STATISTIC
    optional Statistic statistic = 2;

    // must be present when type == TYPE_QUERY_STATISTICTABLE
    optional StatisticTable statisticTable = 3;
  }

  message Update
  {
    // Clear component statistic element values
    message StatisticClear
    {
      // statistic element names to clear
      // specify none to clear all
      repeated string names = 1;
    }

    // Clear component statistic table values
    message StatisticTableClear
    {
      // statistic table names to clear
      // specify none to clear all
      repeated string names = 1;
    }

    // set to the appropriate Update type
    required UpdateMessageType type = 1;
    
    // must be present when type == TYPE_UPDATE_STATISTICCLEAR
    optional StatisticClear statisticClear = 2;

    // must be present when type == TYPE_UPDATE_STATISTICTABLECLEAR
    optional StatisticTableClear statisticTableClear = 3;

  }

  message Error
  {
    enum ErrorType
    {
      TYPE_ERROR_PARAMETER = 1;
      TYPE_ERROR_MALFORMED = 2;
    }

    // error type
    required ErrorType type = 1;

    // error description
    required string description = 2;
  }

  enum ResponseMessageType
  {
    TYPE_RESPONSE_QUERY = 1;
    TYPE_RESPONSE_UPDATE = 2;
    TYPE_RESPONSE_ERROR = 3;
  }

  // message sequence number
  required uint32 sequence = 1;
  
  // reference sequence number corresponding to Request
  // message sequence number
  required uint32 reference = 2;
  
  // set to the appropriate Response message type
  required ResponseMessageType type = 3;

  // must be present when type == TYPE_RESPONSE_QUERY
  optional Query query = 4;

  // must be present when type == TYPE_RESPONSE_UPDATE
  optional Update update = 5;
  
  // must be present when type == TYPE_RESPONSE_ERROR
  optional Error error = 6;
}
